File: Data Structures & Algorithms/unique-paths-ii/submission-2.java
class Solution {
    int [][] visited;
    public int uniquePathsWithObstacles(int[][] grid) {
        visited= new int[grid.length][grid[0].length];
        for(int[] row:visited){
            Arrays.fill(row,-1);
        }
        return rec(0,0,grid);
    }

    public int rec(int i,int j,int[][] grid){
        if(i>=grid.length || i<0 || j<0 || j>=grid[0].length || grid[i][j]==1){
            return 0;
        }
        if(visited[i][j]!=-1){
            // visited[i][j]++;
            return visited[i][j];
        }
        if(i==grid.length-1 && j==grid[0].length-1){
            return 1;
        }
        // visited[i][j]=1;
        int bot = rec(i+1,j,grid);
        int rig = rec(i,j+1,grid);
        visited[i][j]=bot+rig;
        return bot+rig;
    }
}

File: Data Structures & Algorithms/swim-in-rising-water/submission-2.js
class Solution {
    /**
     * @param {number[][]} grid
     * @return {number}
     */
    swimInWater(grid) {
    const minpq = new MinPriorityQueue((a)=>a.val);
    let visited=Array.from({length:grid.length}, ()=>Array(grid.length).fill(0));
    minpq.enqueue({r:0,c:0,val:grid[0][0]});
    let maxx = 0;
    while(minpq.size()!==0){

        const {r,c,val} = minpq.dequeue();
        maxx = Math.max(maxx,val);
        if(r===grid.length-1 && c===grid.length-1){
            break;
        }
       
        let x = [1,-1];
        for(let i of x){
           if(r+i<grid.length && r+i>=0 && visited[r+i][c]===0){
                visited[r+i][c]=1;
                minpq.enqueue({r:r+i,c:c,val:grid[r+i][c]});
           }
           if(c+i<grid.length && c+i>=0 && visited[r][c+i]===0){
                visited[r][c+i]=1;
                minpq.enqueue({r:r,c:c+i,val:grid[r][c+i]});
           }
        }

    }
    return maxx;
}
}
